### PR TITLE
gh-106554: Replace `_BaseSelectorImpl._key_from_fd` with a simple `get`

### PR DESCRIPTION
**Pull request by bdraco**
* Issue: gh-106554

### CODE REVIEW

**Comment by bdraco** [comment]
```python
import timeit
import math
import select
import os
from selectors import EpollSelector, EVENT_WRITE, EVENT_READ


class OriginalEpollSelector(EpollSelector):
    def select(self, timeout=None):
        if timeout is None:
            timeout = -1
        elif timeout <= 0:
            timeout = 0
        else:
            # epoll_wait() has a resolution of 1 millisecond, round away
            # from zero to wait *at least* timeout seconds.
            timeout = math.ceil(timeout * 1e3) * 1e-3
        # epoll_wait() expects `maxevents` to be greater than zero;
        # we want to make sure that `select()` can be called when no
        # FD is registered.
        max_ev = max(len(self._fd_to_key), 1)
        ready = []
        try:
            fd_event_list = self._selector.poll(timeout, max_ev)
        except InterruptedError:
            return ready
        for fd, event in fd_event_list:
            events = 0
            if event & ~select.EPOLLIN:
                events |= EVENT_WRITE
            if event & ~select.EPOLLOUT:
                events |= EVENT_READ

            key = self._key_from_fd(fd)
            if key:
                ready.append((key, events & key.events))
        return ready


class NewEpollSelector(EpollSelector):
    def select(self, timeout=None):
        if timeout is None:
            timeout = -1
        elif timeout <= 0:
            timeout = 0
        else:
            # epoll_wait() has a resolution of 1 millisecond, round away
            # from zero to wait *at least* timeout seconds.
            timeout = math.ceil(timeout * 1e3) * 1e-3
        # epoll_wait() expects `maxevents` to be greater than zero;
        # we want to make sure that `select()` can be called when no
        # FD is registered.
        max_ev = max(len(self._fd_to_key), 1)
        ready = []
        try:
            fd_event_list = self._selector.poll(timeout, max_ev)
        except InterruptedError:
            return ready
        for fd, event in fd_event_list:
            events = 0
            if event & ~select.EPOLLIN:
                events |= EVENT_WRITE
            if event & ~select.EPOLLOUT:
                events |= EVENT_READ

            key = self._fd_to_key.get(fd)
            if key:
                ready.append((key, events & key.events))
        return ready


original_epoll = OriginalEpollSelector()
new_epoll = NewEpollSelector()


for _ in range(512):
    r, w = os.pipe()
    os.write(w, b"a")
    original_epoll.register(r, EVENT_READ)
    new_epoll.register(r, EVENT_READ)


original_time = timeit.timeit(
    "selector.select()",
    number=100000,
    globals={"selector": original_epoll},
)
new_time = timeit.timeit(
    "selector.select()",
    number=100000,
    globals={"selector": new_epoll},
)

print("original: %s" % original_time)
print("new: %s" % new_time)
```

**Comment by bdraco** [comment]
512 pipes
```
original: 9.91608710300352
new: 9.240273159986828
```

1024 pipes
```
original: 20.881464588004746
new: 19.727144259988563
```

**Comment by bdraco** [comment]
A more aggressive change to reduce the repeated calls inside the loop could shave some more time off
```
original: 10.050354178994894
new: 8.222623112000292
```

something like

```python
NOT_EPOLLIN = ~select.EPOLLIN
NOT_EPOLLOUT = ~select.EPOLLOUT

class NewEpollSelector(EpollSelector):
    def select(self, timeout=None):
        if timeout is None:
            timeout = -1
        elif timeout <= 0:
            timeout = 0
        else:
            # epoll_wait() has a resolution of 1 millisecond, round away
            # from zero to wait *at least* timeout seconds.
            timeout = math.ceil(timeout * 1e3) * 1e-3
        # epoll_wait() expects `maxevents` to be greater than zero;
        # we want to make sure that `select()` can be called when no
        # FD is registered.
        max_ev = max(len(self._fd_to_key), 1)
        ready = []
        try:
            fd_event_list = self._selector.poll(timeout, max_ev)
        except InterruptedError:
            return ready

        for fd, event in fd_event_list:
            key = self._fd_to_key.get(fd)
            if key:
                ready.append(
                    (
                        key,
                        (
                            (event & NOT_EPOLLIN and EVENT_WRITE)
                            | (event & NOT_EPOLLOUT and EVENT_READ)
                        )
                        & key.events,
                    )
                )
        return ready
```

**Comment by bdraco** [comment]
`KqueueSelector` could shave off some more time as well with something like
```python
        def select(self, timeout=None):
            timeout = None if timeout is None else max(timeout, 0)
            # If max_ev is 0, kqueue will ignore the timeout. For consistent
            # behavior with the other selector classes, we prevent that here
            # (using max). See https://bugs.python.org/issue29255
            max_ev = max(len(self._fd_to_key), 1)
            ready = []
            try:
                kev_list = self._selector.control(None, max_ev, timeout)
            except InterruptedError:
                return ready
            for kev in kev_list:
                key = self._fd_to_key.get(kev.ident)
                if key:
                    flag = kev.filter
                    if flag == select.KQ_FILTER_READ:
                        events = EVENT_READ
                    elif flag == select.KQ_FILTER_WRITE:
                        events = EVENT_WRITE
                    else:
                        events = 0
                    ready.append((key, events & key.events))
            return ready
```

File: Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Reduce selector overhead
```
Comment (bdraco) [inline Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst]:
```suggestion
Reduce selector overhead by switching `SelectSelector` to use a `dict` `.get()` for `fd`s
```

File: Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Reduce selector overhead by switching `SelectSelector` to use a `dict` `.get()` for `fd`s
```
Comment (bdraco) [inline Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst]:
```suggestion
Reduce selector overhead by switching ``SelectSelector`` to use a ``dict`` ``.get()`` for ``fd``s
```

File: Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Reduce selector overhead by switching ``SelectSelector`` to use a ``dict`` ``.get()`` to fetch file descriptors.
```
Comment (bdraco) [inline Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst]:
```suggestion
Reduce selector overhead by switching ``SelectSelector`` to use a ``dict`` ``.get()`` to lookup file descriptors.
```

**Comment by bdraco** [comment]
With https://github.com/python/cpython/pull/106555#issuecomment-1632025959 the selector overhead falls below the cost of `socket.recv` in the profiles. Its still high on the list at `#4` for my home assistant install, eclipsed only by `socket.recv`, `_run_once` and some bluetooth overhead.

File: Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Reduce selector overhead by switching ``SelectSelector`` to use a ``dict`` ``.get()`` to lookup file descriptors.
```
Comment (methane) [inline Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst]:
```suggestion
:mod:`selectors`: Reduce Selector overhead by using a ``dict.get()`` to lookup file descriptors.
```

Comment (bdraco) [reply Misc/NEWS.d/next/Library/2023-07-09-01-59-24.gh-issue-106554.37c53J.rst]: Thanks

**Comment by bdraco** [comment]
Thanks.  Will work on the follow ups from above

**Pull request merged**